Font joypixels emoji

toonn opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

The JoyPixels emoji font (formerly EmojiOne) was packaged using a CDN endpoint created for Arch Linux.
Arch Linux received an altered license for distribution of the font. Furthermore,  the license requires explicit acceptance from users. And last but not least, as-is the font doesn't work on macOS.

I expect foss licenses to be complied to. In turn, this means we should respect the terms other projects put in their licenses. I contacted the JoyPixels licensing team and they were happy to provide CDN endpoints and similarly altered licenses for distribution of their font for NixOS and nix-darwin.

I altered the expression to work on macOS, which requires downloading and installing a "TrueType Collection" file rather than the TrueType font file. I also updated the licensing information and the package now requires an override to indicate acceptance of the license before you can install it.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

toonn committed:
joypixels: Use dedicated NixOS cdn endpoint

The Arch Linux endpoint was not intended to be used as-is by other
distros. I have asked for and received a proper license with a dedicated
endpoint for NixOS.

toonn committed:
joypixels: Add support for darwin

toonn committed:
joypixels: Refactor src and installPhase for clarity

toonn committed:
joypixels: Update metadata

toonn committed:
joypixels: Add proper license

The JoyPixels font comes with a license which requires explicit
acceptance by the user.

Infinisil commented on pkgs/data/fonts/joypixels/default.nix:
```diff
@@ -1,24 +1,95 @@
+    }.${system} or throwSystem;
+
+    joypixels-free-license = {
+      spdxId = "LicenseRef-JoyPixels-Free-6.0";
```
SPDX is a [license list](https://spdx.org/licenses/), and this license is not part of it, so it doesn't have any SPDX ID. Same for the other `spdxId`s

toonn replied:
That's not true. There's a list of standard license identifiers but SPDX is much more than that. My identifiers are to spec afaict.

toonn replied:
Where to spec means, "LicenseRef-<basically anything goes>" : )
[A reference from the spec.](https://spdx.github.io/spdx-spec/appendix-IV-SPDX-license-expressions/#simple-license-expressions)

Infinisil replied:
Ah I didn't know that, thanks for telling me!

Infinisil commented on pkgs/data/fonts/joypixels/default.nix:
```diff
@@ -1,24 +1,95 @@
+
+in
+
+assert !acceptLicense -> throw ''
```
This is a bit abusive of an `assert`, since it's not really made for this. But I guess it works, and it's pretty simple, so why not

toonn replied:
I'm open to a better way to do this. All I want is a way to visually warn a user and instruct them on how to fix the issue.

jtojnar replied:
Could we at least move it to `src` attribute so that `joypixels.override { acceptLicense = true; }` works?

toonn replied:
I'm fine with this change. Could you explain how passing `joypixels.override { acceptLicense = true; }` differs from specifying it in the overlay like this `super.callPackage joypixels/default.nix { acceptLicense = true; };`.

I expected the overlay to work the same way an override would so I didn't stumble on this.

jtojnar replied:
`makeOverridable` checks the type of the package expression after passing the original arguments:

https://github.com/NixOS/nixpkgs/blob/11af562a96f22dde27d20dc23604bde43964dd2b/lib/customisation.nix#L67-L81

so if the assertion fails you are out of luck.

If you pass the `acceptLicense = true` to the original arguments, then it cannot complain.

toonn replied:
I *could* use some advice on formatting the nix expression. The error message should be formatted for 80 columns imo, but, as it is the indentation causes it to overflow when editing the expression which is a bit confusing.

Should I reformat overflowing urls using string concatenation?

jtojnar replied:
IMO, overflowing lines is still less pain than joining paths. Most editors can visually mark soft-wrapped lines just fine.

toonn commented on pkgs/data/fonts/joypixels/default.nix:
```diff
@@ -1,24 +1,95 @@
-{ stdenv, fetchurl }:
+{ stdenv, fetchurl, acceptLicense ? false }:
```
Are you sure a font warrants a configuration option? nix-env users, and everyone else, would be able to use a simple overlay, no?

Infinisil replied:
`config` values are used very freely in nixpkgs, since they don't cost anything. And setting a `config` value is pretty easy on both NixOS and with `nix-env`, much easier than an overlay. An overlay for `nix-env` to do such an override would be such a file in ~/.config/nixpkgs/overlays/joypixel-accept
```nix
self: super: {
  joypixels = super.joypixels.override { acceptLicense = true; };
}
```

In comparison, with `config`, it's just this in ~/.config/nixpkgs/config.nix:
```nix
{
  joypixels.acceptLicense = true;
}
```

And it's similarly easy on NixOS

toonn replied:
Applied the suggestion 👍 
The way I did it actually allows *both* specifying the config option or overriding the argument so now *everybody*'s happy : )

toonn commented on pkgs/data/fonts/joypixels/default.nix:
```diff
@@ -1,24 +1,95 @@
+
+    systemTag = {
+      x86_64-darwin = "nix-darwin";
+      x86_64-linux = "nix-os";
```
I like this. It's what I wanted but I couldn't come up with a way to do it : )

I'll take another critical look at the variables but IMO the very long almost-duplicate values were less readable.

toonn replied:
I addressed the grouping but haven't removed any variables yet. Are there any that stand out as particularly unnecessary to you?

toonn committed:
joypixels: Refactor to group system-specific variables

toonn committed:
joypixels: Switch from override to config option

Switch from an overridden variable to an option specified in
configuration.nix or config.nix:
  ```
  joypixels.acceptLicense = true;
  ```

jtojnar commented on pkgs/data/fonts/joypixels/default.nix:
```diff
@@ -1,24 +1,90 @@
+        fontFile = "Apple%20Color%20Emoji.ttc";
+        name = "joypixels-apple-color-emoji.ttc";
+      };
+      x86_64-linux = rec {
```
Could this be made generic? There are more and more people who use NixOS on ARM devices (e.g. `aarch64-linux`). Similarly Apple is switching to ARM too.

toonn replied:
How about the following?
```
{ x86_64-darwin = { ... }; }.${system} or { ... } # The NixOS version of the arguments.
```

toonn replied:
Maybe branching on system is clearer? I only used the attrset approach because people had me switch from branches to such attrsets when I made the wire-desktop expression darwin-compatible.

toonn replied:
Went ahead and implemented my suggestion.

toonn replied:
Not sure whether there's a way to specify the system regardless of the architecture?

toonn replied:
Should I just use isDarwin?

Infinisil replied:
You can switch on `stdenv.hostPlatform.parsed.kernel.name` instead. This is `linux` for any Linux and `darwin` for macOS

toonn replied:
👍

toonn committed:
joypixels: Refactor to special-case darwin

Specifying the system-specific variables for x86_64-darwin and
x86_64-linux is too restrictive, excluding for example i686-linux.
Since macOS seems to be the odd one out we can special-case only
x86_64-darwin.

toonn committed:
joypixels: Switch on kernel.name rather than system

x86_64-darwin is too specific because macOS runs on multiple
architectures.

toonn committed:
joypixels: Mark license as unfree

toonn committed:
joypixels: Reword license acceptance warning

toonn committed:
joypixels: Drop dash from nixos endpoint url

The url was changed on request.

toonn committed:
joypixels: Move assert to allow override

By moving the assert concerning license acceptance into the src
attribute license acceptance can be expressed with an override,
`joypixels.override { acceptLicense = true; }`.

toonn committed:
joypixels: Use updated font for macOS

After some back and forth with JoyPixels they agreed to creating a
version of their font for macOS that does not use the exact same name as
the Apple Color Emoji default font.

This naming collision meant it was impossible to configure applications
to use the JoyPixels emoji font unless you disabled the Apple Color
Emoji font using Font Book. Which meant the JoyPixels font could either
replace the Apple Color Emoji font completely or only fill in the gaps
in that font (on my system "hot face" isn't in the system font) but not
be used entirely for specific apps or be used with the system font as a
back up.

toonn commented:
Whoops, been waiting on activity for this. Could've sworn I'd pressed the ready-for-review button before.

jtojnar committed:
joypixels: run install hooks

Also clean up formatting.

jtojnar reviewed (approved):
Seems to work for me, thanks.
